python_modules/dagster-graphql/dagster_graphql/schema/pipelines/pipeline.py

```diff
@@ -258,6 +278,61 @@ def resolve_assetMaterializations(
         )
         return [GrapheneMaterializationEvent(event=event) for event in events]
 
+    def resolve_assetMaterializationHistory(
+        self,
+        graphene_info: ResolveInfo,
+        partitions: Optional[Sequence[str]] = None,
+        partitionInLast: Optional[int] = None,
+        beforeTimestampMillis: Optional[str] = None,
+        afterTimestampMillis: Optional[str] = None,
+        limit: Optional[int] = None,
+        eventTypeSelector: GrapheneMaterializationHistoryEventTypeSelector = GrapheneMaterializationHistoryEventTypeSelector.ALL,
+    ) -> Sequence[GrapheneAssetMaterializationEventType]:
+        from dagster_graphql.implementation.fetch_assets import (
+            get_asset_failed_to_materialize_events,
+            get_asset_materializations,
+        )
+
+        before_timestamp = parse_timestamp(beforeTimestampMillis)
+        after_timestamp = parse_timestamp(afterTimestampMillis)
+        if partitionInLast and self._definition:
+            partitions = self._definition.get_partition_keys()[-int(partitionInLast) :]
+
+        failure_events = []
+        success_events = []
+        if (
+            eventTypeSelector
+            == GrapheneMaterializationHistoryEventTypeSelector.FAILED_TO_MATERIALIZE
+            or eventTypeSelector == GrapheneMaterializationHistoryEventTypeSelector.ALL
+        ):
+            failure_events = [
+                GrapheneFailedToMaterializeEvent(event=event)
+                for event in get_asset_failed_to_materialize_events(
+                    graphene_info,
+                    self.key,
+                    partitions=partitions,
+                    before_timestamp=before_timestamp,
+                    after_timestamp=after_timestamp,
+                    limit=limit,
+                )
+            ]
+        if (
+            eventTypeSelector == GrapheneMaterializationHistoryEventTypeSelector.MATERIALIZATION
+            or eventTypeSelector == GrapheneMaterializationHistoryEventTypeSelector.ALL
+        ):
+            success_events = [
+                GrapheneMaterializationEvent(event=event)
+                for event in get_asset_materializations(
+                    graphene_info,
+                    self.key,
+                    partitions=partitions,
+                    before_timestamp=before_timestamp,
+                    after_timestamp=after_timestamp,
+                    limit=limit,
+                )
+            ]
+        return sorted(failure_events + success_events, key=lambda event: event.timestamp)[:limit]
```

Comment:
can we sort by ID rather than timestamp? I think the former is more reliably monotonically increasing?

Reply:
will do

Reply:
this is a bit trickier than i thought bc we don't have the storage id available at this point. i can update some callsites to make it work though